cmd/maya-apiserver/app/server/backup_endpoint_v1alpha1.go

```diff
@@ -0,0 +1,388 @@
+}
+
+// checkIfBKPPoolNodeDown will check if pool node is running or not
+func checkIfBKPPoolNodeDown(k8sclient *kubernetes.Clientset, bkp *v1alpha1.BackupCStor) bool {
+	var nodeDown = true
+
+	pod := findPodFromCStorID(k8sclient, bkp.Labels["cstorpool.openebs.io/uid"])
+	if pod == nil {
+		glog.Errorf("Failed to find pod for backup:%v", bkp.Name)
+		return nodeDown
+	}
+
+	if pod.Spec.NodeName == "" {
+		return nodeDown
+	}
+
+	node, err := k8sclient.CoreV1().Nodes().Get(pod.Spec.NodeName, v1.GetOptions{})
+	if err != nil {
+		glog.Infof("Failed to fetch node info for backup:%v: %v", bkp.Name, err)
+		return nodeDown
+	}
+	for _, nodestat := range node.Status.Conditions {
+		if nodestat.Type == corev1.NodeReady && nodestat.Status != corev1.ConditionTrue {
+			glog.Infof("Node:%v is not in ready state", node.Name)
+			return nodeDown
+		}
+	}
+	return !nodeDown
+}
+
+// checkIfBKPPoolPodDown will check if pool pod is running or not
+func checkIfBKPPoolPodDown(k8sclient *kubernetes.Clientset, bkp *v1alpha1.BackupCStor) bool {
+	var podDown = true
+
+	pod := findPodFromCStorID(k8sclient, bkp.Labels["cstorpool.openebs.io/uid"])
+	if pod == nil {
+		glog.Errorf("Failed to find pod for backup:%v", bkp.Name)
+		return podDown
+	}
+
+	for _, containerstatus := range pod.Status.ContainerStatuses {
+		if containerstatus.Name == "cstor-pool-mgmt" {
+			return !containerstatus.Ready
+		}
+	}
+
+	return podDown
+}
+
+// findPodFromCStorID will find the Pod having given cstorID
+func findPodFromCStorID(k8sclient *kubernetes.Clientset, cstorID string) *corev1.Pod {
+	cstorPodLabel := "app=cstor-pool"
+	podlistops := v1.ListOptions{
+		LabelSelector: cstorPodLabel,
+	}
+
+	openebsNs := os.Getenv("OPENEBS_NAMESPACE")
+	if openebsNs == "" {
+		glog.Errorf("Failed to fetch operator namespace")
+		return nil
+	}
+
+	podlist, err := k8sclient.CoreV1().Pods(openebsNs).List(podlistops)
+	if err != nil {
+		glog.Errorf("Failed to fetch pod list :%v", err)
+		return nil
+	}
+
+	for _, pod := range podlist.Items {
+		for _, env := range pod.Spec.Containers[0].Env {
+			if env.Name == "OPENEBS_IO_CSTOR_ID" && env.Value == cstorID {
+				return &pod
+			}
+		}
+	}
+	return nil
+}
+
+// findLastBackupStat will find the status of given backup from last-backup
+func findLastBackupStat(clientset internalclientset.Interface, bkp *v1alpha1.BackupCStor) v1alpha1.BackupCStorStatus {
+	lastbkpname := bkp.Spec.BackupName + "-" + bkp.Spec.VolumeName
+	lastbkp, err := clientset.OpenebsV1alpha1().BackupCStorLasts(bkp.Namespace).Get(lastbkpname, v1.GetOptions{})
+	if err != nil {
+		// Unable to fetch the last backup, so we will return fail state
+		glog.Errorf("Failed to fetch last-backup:%s error:%s", lastbkpname, err.Error())
+		return v1alpha1.BKPCStorStatusFailed
+	}
+
+	// let's check if snapname matches with current snapshot name
+	if bkp.Spec.SnapName == lastbkp.Spec.SnapName || bkp.Spec.SnapName == lastbkp.Spec.PrevSnapName {
```

Comment:
why are we checking with prev snapname? It should match with SnapName right?

Reply:
ok